### ISSUE TITLE
`Parameters<...>` fails to resolve parameter type with dynamic type narrowed with `.. exteds keyof ..`

### ISSUE THREAD

**Issue by kallja** [microsoft/TypeScript#52566]
# Bug Report

### 🔎 Search Terms

`parameters type`
`parameters keyof`

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about everything that mentions `parameters` or `keyof`.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKuEHUCWwAWAxArgOwMbCQHtsBnKAXigG8AoKKAM2wC4oAKAQwCcBzVk4FyTYeASgoA+KADdCSACYBuGgF9lNXMQFRUSHqghcA8l3mH02CnSgAeCwDkOAW2gQAHsAjZ5ZANYQQQgZYeGQ0LDwCLQk2a3omRxdWB2cIa1FWOEgwjBx8IlIAbRSXAF1JawB6SqgAUTdIfAh5flRCTAAbeShsQmAoMC5CeUxcaA5LQyGuKAAjTH75QggSazYAOk3uHhJWAAVuVM8uEhssxBRcyIKSYuxEiFKJcXIpKjUaGmqoAClMbWAhCguAMuF8UAA7qgOP0DlwjoZTpspFwIMBMFxSDRQJAoBY4QiThQoASXMdTucchF8lpCgAiJh0p7KIA)

### 💻 Code

```ts
type TypeWithFunctions = {
  fn: (arg: string) => void;
};

const higherOrderFn =
  <FnName extends keyof TypeWithFunctions>(
    fnName: FnName
  ): TypeWithFunctions[FnName] =>
  // Expected: should not produce an error but does
  (...args: Parameters<TypeWithFunctions[FnName]>) => {};
```

### 🙁 Actual behavior

`Parameters<...>` line errors with 
```
Type '(...args: Parameters<TypeWithFunctions[FnName]>) => void' is not assignable to type 'TypeWithFunctions[FnName]'.
  Type '(...args: Parameters<TypeWithFunctions[FnName]>) => void' is not assignable to type '(arg: string) => void'.
    Types of parameters 'args' and 'arg' are incompatible.
      Type '[arg: string]' is not assignable to type 'Parameters<TypeWithFunctions[FnName]>'.(2322)
```

`Parameters<...>` fails to resolve the parameter type tuple.

### 🙂 Expected behavior

`Parameters<...>` should correctly and completely infer the relevant parameter type.

**Comment by RyanCavanaugh** [comment]
There are a few cases here to think about. `TypeWithFunctions` is obviously oversimple; if every property in it had the same signature, you could just write a less-generic version:
```ts
const higherOrderFn = <FnName extends keyof TypeWithFunctions>(fnName: FnName): TypeWithFunctions[keyof TypeWithFunctions] => (...args: Parameters<TypeWithFunctions[keyof TypeWithFunctions]>) => {
  
};
```
and indeed that works. You can also keep the return type generic without loss of functionality:
```ts
type TypeWithFunctions = {
  fn: (arg: string) => void;
  fn2: (arg: number) => void;
};

const higherOrderFn = <FnName extends keyof TypeWithFunctions>(fnName: FnName): TypeWithFunctions[FnName] => (...args: Parameters<TypeWithFunctions[keyof TypeWithFunctions]>) => {
  
};

higherOrderFn("fn")("ok");
higherOrderFn("fn2")(32);
higherOrderFn("fn2")("correct error here");
```

However (considering the non-degenerate case of when `fn2` is in the mix), `Parameters<TypeWithFunctions[FnName]>` as an assignability source gets treated as its possible inhabitants, `[arg: string] | [arg: number]`, and obviously that's not directly correlated to the lookup type.

It's also the case that we can't say for certain that this expression form necessarily assigns to an arbitrary type in the target. For example, `TypeWithFunctions` might well be
```ts
type TypeWithFunctions = {
  fn: (arg: string) => void;
  fn2: (arg: number) => void;
  fn3: {
    (arg: number): void;
    x: string;
  }
};
```
*but* if you're disavowing that possibility, then this example can be rewritten WLOG as
```ts
type TypeWithFunctions = {
  fn: [arg: string];
  fn2: [arg: number];
};

const higherOrderFn = <FnName extends keyof TypeWithFunctions>(fnName: FnName): (...args: TypeWithFunctions[FnName]) => void => (...args: TypeWithFunctions[FnName]) => {
  
};
```

So anyway there's not a steadfast argument for allowing this assignment, in a place that matters, unless TS was going to do wholesale enumeration of all the possible values of `FnName`, which doesn't really scale.

So on net the logic here to make this "work" doesn't scale up, and you don't really need it in the first place since the slightly-less-generic form works just as well (including in the function body).
